### PR TITLE
[4.x] Fields - Empty trash

### PR DESCRIPTION
**Pull request by brianteeman**
Everywhere that we have an empty trash button it asks you to confirm before the trash is emptied as this is a destructive change.

Fields was missing this.

### Testing Instructions
Create and trash a field
Filter the fields by status trashed
select the field
click on empty trash inn the toolbar

### Actual result BEFORE applying this Pull Request
fields are immediately deleted

### Expected result AFTER applying this Pull Request
popup message is displayed to confirm before deletion

### CODE REVIEW

**Comment by richard67** [comment]
@brianteeman The system tests are failing for this PR, and I think it could be related to your changes so the tests might need some changes, too. See the log here https://ci.joomla.org/joomla/joomla-cms/53603/1/30 :

```
 I delete field "Field title","Field deleted"
   I am on page "/administrator/index.php?option=com_fie..."
  [GET] http://localhost/mysql/administrator/index.php?option=com_fields&view=fields&context=com_content.article
   I get config "timeout"
   I wait for element {"id":"filter_search"},90
   I execute js "[].forEach.call(document.querySelectorA..."
   I select option "//*[@id="filter_state"]","-2"
   I wait 2
   I fill field {"id":"filter_search"},"Field title"
   I click {"id":"filter_search"}
   I check all results 
  Selecting Checkall button
   I wait 2
   I click toolbar button "Empty trash"
   I get config "timeout"
   I wait for text "Field deleted",90,{"id":"system-mess...}
  Screenshot and page source were saved into '/********/src/tests/Codeception/_output/' dir
 ERROR 
```

The screenshot then looks like this: https://ci.joomla.org/artifacts/joomla/joomla-cms/4.1-dev/37650/system-tests/53603/ContentFieldListCest.Field.mysql.fail.png

**Comment by richard67** [comment]
Will try to get someone to advise with the system tests.

**Comment by laoneo** [comment]
The system test needs to be adapted of fields. Something like here shuould probably work https://github.com/joomla/joomla-cms/blob/4.1-dev/tests/Codeception/_support/Step/Acceptance/Administrator/Content.php#L229.

**Comment by brianteeman** [comment]
will take a look this afternoon

**Comment by richard67** [comment]
@brianteeman I think it needs to replace this line here 

https://github.com/joomla/joomla-cms/blob/4.1-dev/tests/Codeception/_support/Step/Acceptance/Administrator/Field.php#L121

by these 2 lines:

```
$I->wait(2);
$I->acceptPopup();
```

**Comment by richard67** [comment]
It seems to work.

File: tests/Codeception/_support/Step/Acceptance/Administrator/Field.php [hunk 1]
```diff
@@ -118,6 +118,7 @@ public function deleteField($title, $message)
 		$I->checkAllResults();
 		$I->wait(2);
 		$I->clickToolbarButton('Empty trash');
-		$I->assertSuccessMessage($message);
```
Comment (laoneo) [inline tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: Please do not remove the assert. This is needed as it actually tests if the message is correct. You should just add the accept popup code.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: tests/Codeception/_support/Step/Acceptance/Administrator/Field.php (hunk 1 above)
Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: > You should just add the accept popup code.

@laoneo Before or after the `$I->assertSuccessMessage($message);`?

Comment (laoneo) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: Before

Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: But here we do not have the assert: https://github.com/joomla/joomla-cms/blob/4.1-dev/tests/Codeception/_support/Step/Acceptance/Administrator/Content.php#L229

Comment (laoneo) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: Because this implementation of a step is wrong. A step should just have reusable code which can be used in different tests to eliminate duplicate code, but should not do any assert. So this needs to be fixed in a different pr. For now just add the wait and popup before the assert.

Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: @laoneo Should there maybe be another wait between the acceptPopup and the assert?

Comment (laoneo) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: Waits should be avoided whenever possible. Mostly Codeception waits till either the element appears or it runs in into a timeout. In this case it waits for the text here https://github.com/joomla/joomla-cms/blob/4.1-dev/tests/Codeception/_support/Step/Acceptance/Administrator/Field.php#L57-L61

Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: I see. Thanks.

**Comment by brianteeman** [comment]
This PR brings the behaviour of Empty Trash in Fields to be consistent with all other uses. The test has also been updated to be consistent with all the other tests for Empty Trash.

If the test is incomplete then it is incomplete eveyrwhere and should be updated in its own pull request.

**Comment by richard67** [comment]
> This PR brings the behaviour of Empty Trash in Fields to be consistent with all other uses. The test has also been updated to be consistent with all the other tests for Empty Trash.
> 
> If the test is incomplete then it is incomplete eveyrwhere and should be updated in its own pull request.

I would be ok with that.

File: tests/Codeception/_support/Step/Acceptance/Administrator/Field.php (hunk 1 above)
Comment (laoneo) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: This is not resolved, please add again `$I->assertSuccessMessage($message);` code.

Comment (brianteeman) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:121]: Please make up your mind.

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 36524df7974444596b9a4c16645db4936ca0e6d6<br /><br />Tested successfully in Joomla 4.1.3-dev of 25 April.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37650">issues.joomla.org/tracker/joomla-cms/37650</a>.</sub>

**Comment by beefcakefu** [comment]
I have tested this item :white_check_mark: successfully on 36524df7974444596b9a4c16645db4936ca0e6d6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37650">issues.joomla.org/tracker/joomla-cms/37650</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37650">issues.joomla.org/tracker/joomla-cms/37650</a>.</sub>

File: tests/Codeception/_support/Step/Acceptance/Administrator/Field.php [hunk 2]
```diff
@@ -118,6 +118,7 @@ public function deleteField($title, $message)
...
+		$I->wait(2);
+		$I->acceptPopup();
```
Comment (richard67) [inline tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:122]:
```suggestion
		$I->acceptPopup();
		$I->assertSuccessMessage($message);
```
@brianteeman Could you apply this change so the old behaviour of this test is restored? Otherwise, without that line, the `$message` parameter would be useless here: https://github.com/joomla/joomla-cms/blob/36524df7974444596b9a4c16645db4936ca0e6d6/tests/Codeception/_support/Step/Acceptance/Administrator/Field.php#L102

I agree with you that we should check the same method at other places and if necessary fix them, too.

Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:122]: P.S.: It will not need new human tests after this change will be applied.

Comment (brianteeman) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:122]: i will commit the suggestion as yu asked nicely. i still believe it to be wrong to do it in this pr but ...

Comment (richard67) [reply tests/Codeception/_support/Step/Acceptance/Administrator/Field.php:122]: Thanks. Much appreciated. I agree with you at least to a large part but wanted to avoid arguments with other maintainers 😉 .

**Comment by richard67** [comment]
@laoneo Now as your suggested change has been implemented, please approve the changes on GitHub so the it does not show any requested changes anymore. Thanks in advance.

**Comment by richard67** [comment]
Previous human test results and RTC status are still valid since the commit after that only changed the system tests, which are still passing.

**Review by laoneo — APPROVED** [review]

**Comment by bembelimen** [comment]
Thx

**Comment by brianteeman** [comment]
thanks
